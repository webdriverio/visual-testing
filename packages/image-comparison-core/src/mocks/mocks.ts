import type { BeforeScreenshotOptions } from '../helpers/beforeScreenshot.interfaces.js'
import type { InternalSaveMethodOptions } from '../commands/save.interfaces.js'

export const BEFORE_SCREENSHOT_OPTIONS: BeforeScreenshotOptions = {
    instanceData: {
        appName: 'chrome-app',
        browserName: 'chrome',
        browserVersion: '75.0.1',
        deviceName: '',
        devicePixelRatio: 1,
        initialDevicePixelRatio: 1,
        deviceRectangles: {
            statusBar: { y: 0, x: 0, width: 0, height: 0 },
            homeBar: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 0, width: 0 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
            viewport: { y: 0, x: 0, width: 0, height: 0 },
            bottomBar: { y: 0, x: 0, width: 0, height: 0 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
        },
        isAndroid: false,
        isIOS: false,
        isMobile: false,
        logName: 'chrome-latest',
        name: 'chrome-name',
        nativeWebScreenshot: true,
        platformName: 'Windows 10',
        platformVersion: '1234',
    },
    addressBarShadowPadding: 6,
    disableBlinkingCursor: true,
    disableCSSAnimation: true,
    enableLayoutTesting: false,
    noScrollBars: true,
    toolBarShadowPadding: 6,
    hideElements: [<HTMLElement>(<any>'<div></div>')],
    removeElements: [<HTMLElement>(<any>'<div></div>')],
    waitForFontsLoaded: true,
}
export const CONFIGURABLE = {
    writable: true,
    configurable: true,
}
export const NAVIGATOR_APP_VERSIONS = {
    ANDROID: {
        7: '5.0 (Linux; Android 7.1.1; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        8: '5.0 (Linux; Android 8.1; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        9: '5.0 (Linux; Android 9; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        10: '5.0 (Linux; Android 10; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        11: '5.0 (Linux; Android 11; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
    },
    IOS: {
        10: '5.0 (iPhone; CPU iPhone OS 10_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/10.0 Mobile/15E148 Safari/604.1',
        11: '5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1',
        12: '5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1',
        13: '5.0 (iPhone; CPU iPhone OS 13_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Mobile/15E148 Safari/604.1',
        14: '5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1',
        15: '5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1',
    },
    IPADOS: {
        13: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.1 Safari/605.1.15',
        14: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Safari/605.1.15',
        15: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Safari/605.1.15',
    },
}
export const ANDROID_DEVICES = {
    NEXUS_5X: {
        height: 732,
        width: 412,
        innerHeight: 604,
        innerWidth: 412,
    },
    NEXUS_5X_INNER_HEIGHT: {
        height: 732,
        width: 412,
        innerHeight: 800,
        innerWidth: 412,
    },
    TABLET_WIDTH: {
        height: 768,
        width: 1024,
        innerHeight: 604,
        innerWidth: 412,
    },
}
export const IOS_DEVICES = {
    IPHONE: {
        height: 667,
        width: 375,
        innerHeight: 553,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPHONE_X: {
        height: 812,
        width: 375,
        innerHeight: 635,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPHONE_HEIGHT: {
        height: 896,
        width: 1024,
        innerHeight: 719,
        innerWidth: 414,
        scrollWidth: 414,
        sideBar: 0,
    },
    IPHONE_11: {
        height: 896,
        width: 375,
        innerHeight: 635,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPAD: {
        height: 1366,
        width: 1024,
        innerHeight: 1292,
        innerWidth: 1024,
        scrollWidth: 1024,
        sideBar: 0,
    },
    IPAD_LANDSCAPE: {
        height: 1366,
        width: 1024,
        innerHeight: 746,
        innerWidth: 1046,
        scrollWidth: 1046,
        sideBar: 320,
    },
    IPAD_BIG_SIZE: {
        height: 5432,
        width: 9876,
        innerHeight: 5324,
        innerWidth: 9768,
        scrollWidth: 9768,
        sideBar: 108,
    },
    IPAD_PRO_LANDSCAPE: {
        height: 1366,
        width: 1024,
        innerHeight: 954,
        innerWidth: 1046,
        scrollWidth: 1046,
        sideBar: 320,
    },
}
export const BASE_CHECK_OPTIONS = {
    wic: {
        addressBarShadowPadding: 6,
        autoElementScroll: true,
        addIOSBezelCorners: false,
        autoSaveBaseline: false,
        clearFolder: false,
        userBasedFullPageScreenshot: false,
        enableLegacyScreenshotMethod: false,
        formatImageName: '{tag}-{logName}-{width}x{height}-dpr-{dpr}',
        isHybridApp: false,
        savePerInstance: true,
        toolBarShadowPadding: 6,
        disableBlinkingCursor: false,
        disableCSSAnimation: false,
        enableLayoutTesting: false,
        fullPageScrollTimeout: 1500,
        hideScrollBars: true,
        waitForFontsLoaded: true,
        compareOptions: {
            ignoreAlpha: false,
            ignoreAntialiasing: false,
            ignoreColors: false,
            ignoreLess: false,
            ignoreNothing: false,
            rawMisMatchPercentage: false,
            returnAllCompareData: false,
            saveAboveTolerance: 0,
            scaleImagesToSameSize: false,
            blockOutSideBar: false,
            blockOutStatusBar: false,
            blockOutToolBar: false,
            createJsonReportFiles: false,
            diffPixelBoundingBoxProximity: 5,
        },
        tabbableOptions: {
            circle: {
                backgroundColor: 'rgba(255, 0, 0, 0.4)',
                borderColor: 'rgba(255, 0, 0, 1)',
                borderWidth: 1,
                fontColor: 'rgba(0, 0, 0, 1)',
                fontFamily: 'Arial',
                fontSize: 10,
                size: 10,
            },
            line: {
                color: 'rgba(255, 0, 0, 1)',
                width: 1,
            },
        }
    },
    instanceData: {
        appName: 'TestApp',
        browserName: 'Chrome',
        browserVersion: '118.0.0.0',
        deviceName: 'iPhone 14',
        devicePixelRatio: 2,
        deviceRectangles: {
            bottomBar: { y: 800, x: 0, width: 390, height: 0 },
            homeBar: { x: 0, y: 780, width: 390, height: 34 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 844, width: 390 },
            statusBar: { x: 0, y: 0, width: 390, height: 47 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 390, height: 47 },
            viewport: { y: 47, x: 0, width: 390, height: 733 }
        },
        initialDevicePixelRatio: 2,
        isAndroid: false,
        isIOS: true,
        isMobile: true,
        logName: 'test-log',
        name: 'test-device',
        nativeWebScreenshot: false,
        platformName: 'iOS',
        platformVersion: '17.0'
    },
    folders: {
        actualFolder: '/test/actual',
        baselineFolder: '/test/baseline',
        diffFolder: '/test/diff'
    },
    testContext: {
        commandName: 'checkScreen',
        framework: 'vitest',
        parent: 'test suite',
        title: 'test title',
        tag: 'test-tag',
        instanceData: {
            browser: { name: 'Chrome', version: '118.0.0.0' },
            deviceName: 'iPhone 14',
            platform: { name: 'iOS', version: '17.0' },
            app: 'TestApp',
            isMobile: true,
            isAndroid: false,
            isIOS: true
        }
    }
}
export const BEFORE_SCREENSHOT_MOCK = {
    browserName: 'chrome',
    browserVersion: '120.0.0',
    deviceName: 'desktop',
    dimensions: {
        body: {
            scrollHeight: 1000,
            offsetHeight: 1000
        },
        html: {
            clientWidth: 1200,
            scrollWidth: 1200,
            clientHeight: 900,
            scrollHeight: 1000,
            offsetHeight: 1000
        },
        window: {
            devicePixelRatio: 2,
            innerHeight: 900,
            innerWidth: 1200,
            isEmulated: false,
            isLandscape: false,
            outerHeight: 1000,
            outerWidth: 1200,
            screenHeight: 1080,
            screenWidth: 1920,
        },
    },
    isAndroid: false,
    isAndroidChromeDriverScreenshot: false,
    isAndroidNativeWebScreenshot: false,
    isIOS: false,
    isMobile: false,
    isTestInBrowser: true,
    isTestInMobileBrowser: false,
    addressBarShadowPadding: 0,
    toolBarShadowPadding: 0,
    appName: '',
    logName: 'chrome',
    name: 'chrome',
    platformName: 'desktop',
    platformVersion: '120.0.0',
    devicePixelRatio: 2,
    deviceRectangles: {
        bottomBar: { height: 0, width: 0, x: 0, y: 0 },
        homeBar: { height: 0, width: 0, x: 0, y: 0 },
        leftSidePadding: { height: 0, width: 0, x: 0, y: 0 },
        rightSidePadding: { height: 0, width: 0, x: 0, y: 0 },
        screenSize: { height: 0, width: 0 },
        statusBar: { height: 0, width: 0, x: 0, y: 0 },
        statusBarAndAddressBar: { height: 0, width: 0, x: 0, y: 0 },
        viewport: { height: 0, width: 0, x: 0, y: 0 }
    },
    initialDevicePixelRatio: 2,
    nativeWebScreenshot: false
}
export const AFTER_SCREENSHOT_MOCK = {
    devicePixelRatio: 2,
    fileName: 'test-screen.png'
}
export const COMMON_METHOD_OPTIONS = {
    disableBlinkingCursor: false,
    disableCSSAnimation: false,
    enableLayoutTesting: false,
    enableLegacyScreenshotMethod: false,
    hideScrollBars: true,
    hideElements: [],
    removeElements: [],
    waitForFontsLoaded: true,
}
export const createBeforeScreenshotMock = (overrides = {}) => ({
    ...BEFORE_SCREENSHOT_MOCK,
    ...overrides
})
export const createAfterScreenshotMock = (overrides = {}) => ({
    ...AFTER_SCREENSHOT_MOCK,
    ...overrides
})
export const createMethodOptions = (overrides = {}) => ({
    ...COMMON_METHOD_OPTIONS,
    ...overrides
})
export const createBaseOptions = (type: 'screen' | 'element', overrides = {}) => {
    const baseOptions = {
        browserInstance: {
            isAndroid: false,
            isMobile: false
        } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: BASE_CHECK_OPTIONS.instanceData,
        tag: `test-${type}`
    }

    if (type === 'screen') {
        return {
            ...baseOptions,
            saveScreenOptions: {
                wic: BASE_CHECK_OPTIONS.wic,
                method: COMMON_METHOD_OPTIONS
            },
            ...overrides
        }
    }

    return {
        ...baseOptions,
        saveElementOptions: {
            wic: BASE_CHECK_OPTIONS.wic,
            method: COMMON_METHOD_OPTIONS
        },
        ...overrides
    }
}

export function createTestOptions<T extends InternalSaveMethodOptions>(
    baseOptions: T,
    overrides: Partial<T> = {}
): T {
    const result = {
        ...baseOptions,
        ...overrides
    } as T

    if ('saveScreenOptions' in baseOptions && 'saveScreenOptions' in result) {
        const baseScreenOptions = baseOptions.saveScreenOptions as any
        const overrideScreenOptions = (overrides as any).saveScreenOptions || {}
        result.saveScreenOptions = {
            ...baseScreenOptions,
            ...overrideScreenOptions,
            wic: {
                ...BASE_CHECK_OPTIONS.wic,
                ...(baseScreenOptions?.wic || {}),
                ...(overrideScreenOptions?.wic || {})
            }
        } as any
    }

    if ('saveElementOptions' in baseOptions && 'saveElementOptions' in result) {
        const baseElementOptions = baseOptions.saveElementOptions as any
        const overrideElementOptions = (overrides as any).saveElementOptions || {}
        result.saveElementOptions = {
            ...baseElementOptions,
            ...overrideElementOptions,
            wic: {
                ...BASE_CHECK_OPTIONS.wic,
                ...(baseElementOptions?.wic || {}),
                ...(overrideElementOptions?.wic || {})
            }
        } as any
    }

    return result
}
