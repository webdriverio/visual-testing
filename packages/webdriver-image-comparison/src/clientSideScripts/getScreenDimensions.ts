import type { ScreenDimensions } from './screenDimensions.interfaces.js'

/**
 * Get all the screen dimensions
 */
export default function getScreenDimensions(isMobile: boolean): ScreenDimensions {
    // We need to determine if the screen is emulated, because that would return different values
    const width = window.innerWidth
    const height = window.innerHeight
    const dpr = window.devicePixelRatio || 1
    const minEdge = Math.min(width, height)
    const maxEdge = Math.max(width, height)
    const isLikelyEmulated =
        !isMobile &&              // Only check for emulated on desktop
        dpr >= 2 &&               // High-DPI signal
        minEdge <= 800 &&         // Catch phones/tablets in portrait/landscape
        maxEdge <= 1280 &&        // Conservative max for emulated tablet sizes
        width > 0 && height > 0   // Sanity check

    // Other checks
    const body = document.body
    const html = document.documentElement
    const bodyDimensions = {
        // On mobile & desktop: Total scrollable height of the body element, including content not visible on screen
        scrollHeight: !body ? 0 : body.scrollHeight,
        // On mobile & desktop: Height of body element including padding but not margin
        offsetHeight: !body ? 0 : body.offsetHeight,
    }
    const htmlDimensions = {
        /** On mobile & desktop: Viewport height excluding scrollbars */
        clientHeight: !html ? 0 : html.clientHeight,
        /** On mobile & desktop: Viewport width excluding scrollbars */
        clientWidth: !html ? 0 : html.clientWidth,
        /** On mobile & desktop: Total scrollable height including overflow */
        scrollHeight: !html ? 0 : html.scrollHeight,
        /** On mobile & desktop: Total scrollable width including overflow */
        scrollWidth: !html ? 0 : html.scrollWidth,
        /** On mobile & desktop: Height including padding and border */
        offsetHeight: !html ? 0 : html.offsetHeight,
    }

    const windowDimensions = {
        /**
         * Mobile: Viewport width (changes with zoom)
         * Desktop: Viewport width including scrollbars
         */
        innerWidth: window.innerWidth,
        /**
         * Mobile: Viewport height (changes with zoom)
         * Desktop: Viewport height including scrollbars
         */
        innerHeight: window.innerHeight,
        /**
         * Mobile: True if device is in landscape orientation
         * Desktop: Based on viewport aspect ratio
         */
        isLandscape: window.matchMedia('(orientation: landscape)').matches,
        /**
         * Mobile: Full browser height including UI elements
         * Desktop: Browser window height including toolbars/status bar
         * Emulated: It will be the same as window.innerHeight
         */
        outerHeight: isLikelyEmulated && window.outerHeight > 0?
            window.innerHeight :
            window.outerHeight === 0 ?
                htmlDimensions.clientHeight :
                window.outerHeight,
        /**
         * Mobile: Full browser width
         * Desktop: Browser window width
         * Emulated: It will be the same as window.innerWidth
         */
        outerWidth: isLikelyEmulated && window.outerWidth > 0 ?
            window.innerWidth :
            window.outerWidth === 0 ?
                htmlDimensions.clientWidth :
                window.outerWidth,
        /**
         * Mobile: Physical pixel ratio (typically >1 for high DPI)
         * Desktop: Usually 1, or 2 for high DPI displays
         */
        devicePixelRatio: window.devicePixelRatio,
        /**
         * Mobile: Always false
         * Desktop: Always false
         * Emulated: Always true
         */
        isEmulated: isLikelyEmulated,
        /**
         * Mobile: Physical screen width in CSS pixels
         * Desktop: Monitor width in pixels
         * Emulated: It will be the same as window.innerWidth
         */
        screenWidth: isLikelyEmulated ? window.innerWidth : window.screen.width,
        /**
         * Mobile: Physical screen height in CSS pixels
         * Desktop: Monitor height in pixels
         * Emulated: It will be the same as window.innerHeight
         */
        screenHeight: isLikelyEmulated ? window.innerHeight : window.screen.height,
    }

    return {
        dimensions: {
            body: bodyDimensions,
            html: htmlDimensions,
            window: windowDimensions,
        },
    }
}
