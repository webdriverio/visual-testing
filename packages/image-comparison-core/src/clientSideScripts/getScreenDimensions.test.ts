// @vitest-environment jsdom

import { describe, it, expect, vi } from 'vitest'
import { CONFIGURABLE } from '../mocks/mocks.js'
import getScreenDimensions from './getScreenDimensions.js'

describe('getScreenDimensions', () => {
    it('should get the needed screen dimensions for a real device', () => {
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: true,
            })),
            ...CONFIGURABLE,
        })
        expect(getScreenDimensions(true)).toMatchSnapshot()
    })

    it('should get the needed screen dimensions', () => {
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: true,
            })),
            ...CONFIGURABLE,
        })
        expect(getScreenDimensions(false)).toMatchSnapshot()
    })

    it('should get the needed screen dimensions if the outerHeight and outerWidth are set to 0', () => {
        Object.defineProperty(window, 'outerHeight', { value: 0 })
        Object.defineProperty(window, 'outerWidth', { value: 0 })
        Object.defineProperty(document.documentElement, 'clientHeight', { value: 1234 })
        Object.defineProperty(document.documentElement, 'clientWidth', { value: 4321 })
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: false,
            })),
            ...CONFIGURABLE,
        })

        expect(getScreenDimensions(false)).toMatchSnapshot()
    })

    it('should return zeroed dimensions if the document attributes are null', () => {
        Object.defineProperty(document, 'body', { value: null })
        Object.defineProperty(document, 'documentElement', { value: null })
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: false,
            })),
            ...CONFIGURABLE,
        })

        expect(getScreenDimensions(false)).toMatchSnapshot()
    })

    it('should detect mobile emulation and return emulated dimensions', () => {
        const mockScreen = {
            width: 375,
            height: 667
        }
        const originalScreen = window.screen

        Object.defineProperty(window, 'screen', {
            value: mockScreen,
            configurable: true,
            writable: true
        })
        Object.defineProperty(window, 'devicePixelRatio', { value: 3, configurable: true })
        Object.defineProperty(window, 'innerWidth', { value: 375, configurable: true })
        Object.defineProperty(window, 'innerHeight', { value: 667, configurable: true })
        Object.defineProperty(window, 'outerWidth', { value: 375, configurable: true })
        Object.defineProperty(window, 'outerHeight', { value: 667, configurable: true })
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: false,
            })),
            ...CONFIGURABLE,
        })

        const dimensions = getScreenDimensions(false)

        Object.defineProperty(window, 'screen', {
            value: originalScreen,
            configurable: true,
            writable: true
        })

        expect(dimensions.dimensions.window.screenWidth).toBe(375)
        expect(dimensions.dimensions.window.screenHeight).toBe(667)
        expect(dimensions.dimensions.window.outerWidth).toBe(375)
        expect(dimensions.dimensions.window.outerHeight).toBe(667)
        expect(dimensions.dimensions.window.devicePixelRatio).toBe(3)
    })

    it('should handle desktop (non-emulated) dimensions correctly', () => {
        const mockScreen = {
            width: 1920,
            height: 1080
        }
        const originalScreen = window.screen

        Object.defineProperty(window, 'screen', {
            value: mockScreen,
            configurable: true,
            writable: true
        })
        Object.defineProperty(window, 'devicePixelRatio', { value: 1, configurable: true })
        Object.defineProperty(window, 'innerWidth', { value: 1440, configurable: true })
        Object.defineProperty(window, 'innerHeight', { value: 900, configurable: true })
        Object.defineProperty(window, 'outerWidth', { value: 1440, configurable: true })
        Object.defineProperty(window, 'outerHeight', { value: 900, configurable: true })
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: true,
            })),
            ...CONFIGURABLE,
        })

        const dimensions = getScreenDimensions(false)

        Object.defineProperty(window, 'screen', {
            value: originalScreen,
            configurable: true,
            writable: true
        })

        expect(dimensions.dimensions.window.screenWidth).toBe(1920)
        expect(dimensions.dimensions.window.screenHeight).toBe(1080)
        expect(dimensions.dimensions.window.outerWidth).toBe(1440)
        expect(dimensions.dimensions.window.outerHeight).toBe(900)
        expect(dimensions.dimensions.window.devicePixelRatio).toBe(1)
    })

    it('should handle high DPI desktop displays', () => {
        const mockScreen = {
            width: 2880,
            height: 1800
        }
        const originalScreen = window.screen

        Object.defineProperty(window, 'screen', {
            value: mockScreen,
            configurable: true,
            writable: true
        })
        Object.defineProperty(window, 'devicePixelRatio', { value: 2, configurable: true })
        Object.defineProperty(window, 'innerWidth', { value: 1440, configurable: true })
        Object.defineProperty(window, 'innerHeight', { value: 900, configurable: true })
        Object.defineProperty(window, 'outerWidth', { value: 1440, configurable: true })
        Object.defineProperty(window, 'outerHeight', { value: 900, configurable: true })
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: true,
            })),
            ...CONFIGURABLE,
        })

        const dimensions = getScreenDimensions(false)

        Object.defineProperty(window, 'screen', {
            value: originalScreen,
            configurable: true,
            writable: true
        })

        expect(dimensions.dimensions.window.devicePixelRatio).toBe(2)
        expect(dimensions.dimensions.window.screenWidth).toBe(2880)
        expect(dimensions.dimensions.window.screenHeight).toBe(1800)
    })

    it('should handle zero devicePixelRatio', () => {
        Object.defineProperty(window, 'devicePixelRatio', { value: 0, configurable: true })
        Object.defineProperty(window, 'innerWidth', { value: 1920, configurable: true })
        Object.defineProperty(window, 'innerHeight', { value: 1080, configurable: true })
        Object.defineProperty(window, 'matchMedia', {
            value: vi.fn().mockImplementation(() => ({
                matches: true,
            })),
            ...CONFIGURABLE,
        })

        const dimensions = getScreenDimensions(false)
        expect(dimensions.dimensions.window.devicePixelRatio).toBe(1)
    })
})
