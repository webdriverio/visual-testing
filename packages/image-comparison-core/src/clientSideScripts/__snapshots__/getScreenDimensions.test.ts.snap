// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getScreenDimensions > should get the needed screen dimensions 1`] = `
{
  "dimensions": {
    "body": {
      "offsetHeight": 0,
      "scrollHeight": 0,
    },
    "html": {
      "clientHeight": 0,
      "clientWidth": 0,
      "offsetHeight": 0,
      "scrollHeight": 0,
      "scrollWidth": 0,
    },
    "window": {
      "devicePixelRatio": 1,
      "innerHeight": 768,
      "innerWidth": 1024,
      "isEmulated": false,
      "isLandscape": true,
      "outerHeight": 768,
      "outerWidth": 1024,
      "screenHeight": 0,
      "screenWidth": 0,
    },
  },
}
`;

exports[`getScreenDimensions > should get the needed screen dimensions for a real device 1`] = `
{
  "dimensions": {
    "body": {
      "offsetHeight": 0,
      "scrollHeight": 0,
    },
    "html": {
      "clientHeight": 0,
      "clientWidth": 0,
      "offsetHeight": 0,
      "scrollHeight": 0,
      "scrollWidth": 0,
    },
    "window": {
      "devicePixelRatio": 1,
      "innerHeight": 768,
      "innerWidth": 1024,
      "isEmulated": false,
      "isLandscape": true,
      "outerHeight": 768,
      "outerWidth": 1024,
      "screenHeight": 0,
      "screenWidth": 0,
    },
  },
}
`;

exports[`getScreenDimensions > should get the needed screen dimensions if the outerHeight and outerWidth are set to 0 1`] = `
{
  "dimensions": {
    "body": {
      "offsetHeight": 0,
      "scrollHeight": 0,
    },
    "html": {
      "clientHeight": 1234,
      "clientWidth": 4321,
      "offsetHeight": 0,
      "scrollHeight": 0,
      "scrollWidth": 0,
    },
    "window": {
      "devicePixelRatio": 1,
      "innerHeight": 768,
      "innerWidth": 1024,
      "isEmulated": false,
      "isLandscape": false,
      "outerHeight": 1234,
      "outerWidth": 4321,
      "screenHeight": 0,
      "screenWidth": 0,
    },
  },
}
`;

exports[`getScreenDimensions > should return zeroed dimensions if the document attributes are null 1`] = `
{
  "dimensions": {
    "body": {
      "offsetHeight": 0,
      "scrollHeight": 0,
    },
    "html": {
      "clientHeight": 0,
      "clientWidth": 0,
      "offsetHeight": 0,
      "scrollHeight": 0,
      "scrollWidth": 0,
    },
    "window": {
      "devicePixelRatio": 1,
      "innerHeight": 768,
      "innerWidth": 1024,
      "isEmulated": false,
      "isLandscape": false,
      "outerHeight": 0,
      "outerWidth": 0,
      "screenHeight": 0,
      "screenWidth": 0,
    },
  },
}
`;
